pkg/custom-plugins/spec_component.go

```diff
@@ -0,0 +1,211 @@
+
+func (c *component) Start() error {
+	log.Logger.Infow("starting custom plugin", "plugin", c.spec.PluginName)
+
+	itv := c.spec.Interval.Duration
+	// either periodic check is disabled or interval is too short
+	if itv < time.Second {
+		_ = c.Check()
+		return nil
+	}
+
+	go func() {
+		ticker := time.NewTicker(itv)
+		defer ticker.Stop()
+
+		for {
+			_ = c.Check()
+
+			select {
+			case <-c.ctx.Done():
+				return
+			case <-ticker.C:
+			}
+		}
+	}()
+	return nil
+}
+
+func (c *component) Check() components.CheckResult {
+	log.Logger.Infow("checking custom plugin", "plugin", c.spec.PluginName)
+
+	cr := &checkResult{
+		ts: time.Now().UTC(),
+	}
+	defer func() {
+		c.lastMu.Lock()
+		c.lastCheckResult = cr
+		c.lastMu.Unlock()
+	}()
+
+	if c.spec.StatePlugin == nil {
+		cr.health = apiv1.HealthStateTypeHealthy
+		cr.reason = "no state plugin defined"
+		return cr
+	}
+
+	cctx, ccancel := context.WithTimeout(c.ctx, c.spec.Timeout.Duration)
+	defer ccancel()
+
+	cr.Output, cr.ExitCode, cr.err = c.spec.StatePlugin.executeAllSteps(cctx)
+
+	if cr.err != nil {
+		cr.health = apiv1.HealthStateTypeUnhealthy
+		cr.reason = fmt.Sprintf("error executing state plugin -- %s", cr.err)
+		return cr
+	}
+
+	cr.health = apiv1.HealthStateTypeHealthy
+	cr.reason = string(cr.Output)
+	return cr
+}
+
+func (c *component) LastHealthStates() apiv1.HealthStates {
+	c.lastMu.RLock()
+	lastCheckResult := c.lastCheckResult
+	c.lastMu.RUnlock()
+	return lastCheckResult.getLastHealthStates()
+}
+
+func (c *component) Events(ctx context.Context, since time.Time) (apiv1.Events, error) {
+	return nil, nil
+}
+
+func (c *component) Close() error {
+	log.Logger.Debugw("closing component")
+
+	c.cancel()
+
+	return nil
+}
+
+var _ components.CheckResult = &checkResult{}
+
+type checkResult struct {
+	Output   []byte
+	ExitCode int32
+
+	name string
+
+	// timestamp of the last check
+	ts time.Time
+	// error from the last check
+	err error
+
+	// tracks the healthy evaluation result of the last check
+	health apiv1.HealthStateType
+	// tracks the reason of the last check
+	reason string
+}
+
+func (cr *checkResult) String() string {
+	if cr == nil {
+		return ""
+	}
+
+	return string(cr.Output) + "\n\n" + fmt.Sprintf("(exit code %d)", cr.ExitCode)
+}
+
+func (cr *checkResult) Summary() string {
+	if cr == nil {
+		return ""
+	}
+	return cr.reason
+}
+
+func (cr *checkResult) HealthState() apiv1.HealthStateType {
+	if cr == nil {
+		return ""
+	}
+	return cr.health
+}
+
+func (cr *checkResult) getError() string {
+	if cr == nil || cr.err == nil {
+		return ""
+	}
+	return cr.err.Error()
+}
+
+func (cr *checkResult) getLastHealthStates() apiv1.HealthStates {
+	if cr == nil {
+		return apiv1.HealthStates{
+			{
+				Component: "custom-plugin",
+				Name:      "custom-plugin",
+				Health:    apiv1.HealthStateTypeHealthy,
+				Reason:    "no data yet",
+			},
+		}
+	}
+
+	state := apiv1.HealthState{
+		Component: "custom-plugin",
+		Name:      cr.name,
```

Comment:
I didn't find where the field is populated

Reply:
fixed